ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -75,132 +76,133 @@ public static ParsePayload of(String schemasLocation) {
-        PerDocTypeCounter.inc(attributes, "error_schema_not_found");
-        PerDocTypeCounter.inc(attributes, "error_submission_bytes", submissionBytes);
-        throw e;
-      }
 
-      try {
-        validateTimed(schema, json);
-      } catch (ValidationException e) {
-        PerDocTypeCounter.inc(attributes, "error_schema_validation");
-        PerDocTypeCounter.inc(attributes, "error_submission_bytes", submissionBytes);
-        throw e;
-      }
-
-      if (metadataStore == null) {
-        metadataStore = PipelineMetadataStore.of(schemasLocation, BeamFileInputStream::open);
-      }
+        try {
+          validateTimed(schema, json);
+        } catch (ValidationException e) {
+          PerDocTypeCounter.inc(attributes, "error_schema_validation");
+          PerDocTypeCounter.inc(attributes, "error_submission_bytes", submissionBytes);
+          throw e;
+        }
 
-      try {
-        deprecationCheck(attributes);
-      } catch (DeprecatedMessageException e) {
-        PerDocTypeCounter.inc(attributes, "error_deprecated_message");
-        PerDocTypeCounter.inc(attributes, "error_submission_bytes", submissionBytes);
-        throw e;
-      }
+        try {
+          deprecationCheck(attributes);
+        } catch (DeprecatedMessageException e) {
+          PerDocTypeCounter.inc(attributes, "error_deprecated_message");
+          PerDocTypeCounter.inc(attributes, "error_submission_bytes", submissionBytes);
+          throw e;
+        }
 
-      addAttributesFromPayload(attributes, json);
+        addAttributesFromPayload(attributes, json);
 
-      // https://github.com/mozilla/gcp-ingestion/issues/780
-      // We need to be careful to consistently use our util methods (which use Jackson) for
-      // serializing and de-serializing JSON to reduce the possibility of introducing encoding
-      // issues. We previously called json.toString().getBytes() here without specifying a
-      // charset.
-      byte[] normalizedPayload = Json.asBytes(json);
+        // https://github.com/mozilla/gcp-ingestion/issues/780
+        // We need to be careful to consistently use our util methods (which use Jackson) for
+        // serializing and de-serializing JSON to reduce the possibility of introducing encoding
+        // issues. We previously called json.toString().getBytes() here without specifying a
+        // charset.
+        byte[] normalizedPayload = Json.asBytes(json);
 
-      PerDocTypeCounter.inc(attributes, "valid_submission");
-      PerDocTypeCounter.inc(attributes, "valid_submission_bytes", submissionBytes);
+        PerDocTypeCounter.inc(attributes, "valid_submission");
+        PerDocTypeCounter.inc(attributes, "valid_submission_bytes", submissionBytes);
 
-      return Collections.singletonList(new PubsubMessage(normalizedPayload, attributes));
+        out.get(outputTag).output(new PubsubMessage(normalizedPayload, attributes));
+      } catch (IOException | SchemaNotFoundException | ValidationException
```

Comment:
It might be worth checking later if using exceptions for control flow here doesn't negatively affect performance.